server/src/main/java/org/elasticsearch/transport/TransportService.java

```diff
@@ -86,6 +89,7 @@ public class TransportService extends AbstractLifecycleComponent
     private final boolean remoteClusterClient;
     private final Transport.ResponseHandlers responseHandlers;
     private final TransportInterceptor interceptor;
+    private final ConcurrentMap<DiscoveryNode, TransportVersion> transportVersions = ConcurrentCollections.newConcurrentMap();
```

Comment:
I am wondering how will this work now. Will all nodes have the same content of this map?
If I understand correctly, at the moment when a new node joins, the `DiscoveryNodes` (plural) object will be updated ina cluster state. So information about nodes/versions will be available to all nodes in the cluster

With this approach on what node will the `onConnectionOpened` be executed? On all nodes? Or just master node?

Reply:
Hmm, yes. Whereas the node version is stored in cluster state, the connection transport version is relative to each node. So two nodes with different versions could negotiate different transport versions to a third node.

However, if each node has got a connection to each other node, they should all see the same minimum transport version, I believe.